pkg/operation/botanist/controlplane/kubecontrollermanager/kube_controller_manager.go

```diff
@@ -0,0 +1,536 @@
+	"github.com/gardener/gardener/pkg/operation/common"
+	"github.com/gardener/gardener/pkg/utils"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+
+	"github.com/Masterminds/semver"
+	resourcesv1alpha1 "github.com/gardener/gardener-resource-manager/pkg/apis/resources/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// ServiceName is the name of the service of the kube-controller-manager.
+	ServiceName = "kube-controller-manager"
+	// kubeControllerManagerContainerName is the name of the kube-controller-manager container
+	kubeControllerManagerContainerName = v1beta1constants.DeploymentNameKubeControllerManager
+
+	// SecretName is a constant for the secret name for the kube-controller-manager's kubeconfig secret.
+	SecretName = "kube-controller-manager"
+	// SecretNameServer is the name of the kube-controller-managerr server certificate secret.
+	SecretNameServer = "kube-controller-manager-server"
+
+	// labelRole is a constant for the value of a label with key 'role' whose value is 'controller-manager'.
+	labelRole = "controller-manager"
+	// portNameMetrics is a constant for the name of the metrics port of the kube-controller-manager.
+	portNameMetrics = "metrics"
+
+	// volumeMountPathCA is the volume mount path for the CA certificate used by the kube controller manager.
+	volumeMountPathCA = "/srv/kubernetes/ca"
+	// volumeMountPathServiceAccountKey is the volume mount path for the service account key that is a PEM-encoded private RSA or ECDSA key used to sign service account tokens.
+	volumeMountPathServiceAccountKey = "/srv/kubernetes/service-account-key"
+	// volumeMountPathKubeconfig is the volume mount path for the kubeconfig which can be used by the kube-controller-manager to communicate with the kube-apiserver.
+	volumeMountPathKubeconfig = "/var/lib/kube-controller-manager"
+	// volumeMountPathServer is the volume mount path for the x509 TLS server certificate and key for the HTTPS server inside the kube-controller-manager (which is used for metrics and health checks).
+	volumeMountPathServer = "/var/lib/kube-controller-manager-server"
+
+	// managedResourceName is the name of the managed resource that contains the resources to be deployed into the Shoot cluster.
+	managedResourceName = "shoot-core-kube-controller-manager"
+)
+
+// KubeControllerManager contains functions for a kube-controller-manager deployer.
+type KubeControllerManager interface {
```

Comment:
I guess there is no strong reason for it why we do it this way. We could also define the interfaces in the `botanist` package (consumer), but I don't see a clear benefit now from doing it this way. Do you? For now, it probably "cleaner" to have it in the same package, not sure. What do other folks think?

> Also I am not sure if an interface in this case is even necessary. Why not reference the concrete type?

In the future we might have mocks for it when we test higher-level functions.

Reply:
@rfranzke WDYT of the comments here: https://github.com/golang/go/wiki/CodeReviewComments#interfaces

> Do not define interfaces on the implementor side of an API "for mocking"; instead, design the API so that it can be tested using the public API of the real implementation.

Seems to be the best practice in go:
https://stackoverflow.com/questions/55105509/golang-interfaces-mocking

Reply:
I don't agree to this statement in general. Mocks will make testing higher-level function (like the shoot flow) much easier. If you don't mock the KCM then you have to duplicate its behaviour in the shoot flow test again (although you already have tests in the KCM package).

Reply:
But I agree with you that it is convenient for mocks.